libp2p/switch.nim

```diff
@@ -417,30 +426,48 @@ proc mount*[T: LPProtocol](s: Switch, proto: T, matcher: Matcher = nil) {.gcsafe
   s.ms.addHandler(proto.codecs, proto, matcher)
 
 proc accept(s: Switch, transport: Transport) {.async.} = # noraises
-  ## transport's accept loop
+  ## switch accept loop, ran for every transport
   ##
 
+  let upgrades = AsyncSemaphore.init(ConcurrentUpgrades)
   while transport.running:
     var conn: Connection
     try:
       debug "About to accept incoming connection"
+      await upgrades.acquire()
       conn = await transport.accept()
-      if not isNil(conn):
-        debug "Accepted an incoming connection", conn
-        asyncSpawn s.upgradeIncoming(conn) # perform upgrade on incoming connection
-      else:
+      if isNil(conn):
         # A nil connection means that we might have hit a
         # file-handle limit (or another non-fatal error),
         # we can get one on the next try, but we should
         # be careful to not end up in a thigh loop that
         # will starve the main event loop, thus we sleep
         # here before retrying.
+        trace "Unable to get a connection, sleeping"
         await sleepAsync(100.millis) # TODO: should be configurable?
```

Comment:
try/catch because may be cancelled, in which case semaphore must be released - no finally in loop though

Reply:
Hmm, isn't canceling only possible on the entire `accept`?

Reply:
yeah, but you don't want to leave the semaphore hanging, ever, same as you don't want to leave other unfinished futures around

Reply:
Those are canceled as well tho? There is (should be?) teardown process where all enclosed futures get cancelled as well.

Reply:
I meant that the semaphore will not be released - it may be harmless in this particular case but it's a ticking bomb not to release the semaphore because the code will get copy-pasted and refactored to places where it matters - the point is that with semaphores and locks, you never ever want to leave them in an acquired state no matter the flow - I mentioned the futures because they're similar from that point of view

Reply:
yeah, I guess it's a good idea to release on cancellation either way